File: src/main/java/com/financemanager/controllers/CreateManagerController.java
package com.financemanager.controllers;

import com.financemanager.IO.importers.BankStatementParser;
import com.financemanager.IO.importers.CGDParser;
import com.financemanager.IO.importers.CreditoAgricolaParser;
import com.financemanager.model.Transaction;
import com.financemanager.service.AccountManager;
import javafx.collections.FXCollections;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.stage.FileChooser;
import javafx.stage.Modality;
import javafx.stage.Stage;

import java.io.File;
import java.io.IOException;
import java.util.List;

public class CreateManagerController {

    @FXML private Button back_to_menu_button;
    @FXML private Button import_file_path_button;
    @FXML private Button import_file_button;
    @FXML private Label file_name_label;

    private String filePath = "";

    @FXML
    public void handleBacktoMenu(ActionEvent event) {
        try {
            FXMLLoader loader = new FXMLLoader(getClass().getResource("/Menu.fxml"));
            Parent root = loader.load();
            Stage stage = (Stage) ((Node) event.getSource()).getScene().getWindow();
            stage.setScene(new Scene(root));
            stage.setTitle("Menu");
            stage.show();
        } catch (IOException e) {
            System.out.println("Error loading FXML: " + e.getMessage());
        }
    }

    @FXML
    public void handleImportMenu(ActionEvent event) {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Import New Manager");
        fileChooser.getExtensionFilters().addAll(
                new FileChooser.ExtensionFilter("CSV/Excel Files", "*.csv", "*.xlsx"),
                new FileChooser.ExtensionFilter("All Files", "*.*")
        );

        Stage currentStage = (Stage) ((Node) event.getSource()).getScene().getWindow();
        File selectedFile = fileChooser.showOpenDialog(currentStage);

        if (selectedFile != null) {
            filePath = selectedFile.getAbsolutePath(); // só aqui, depois do null check
            file_name_label.setText(selectedFile.getName());
            System.out.println("Selected File: " + selectedFile.getAbsolutePath());
        } else {
            System.out.println("Selection Cancelled");
        }
    }

    @FXML
    public void handleImportManager(ActionEvent event) {
        if (filePath.isEmpty()) {
            Alert alert = new Alert(Alert.AlertType.ERROR);
            alert.setTitle("Selection Error");
            alert.setHeaderText(null);
            alert.setContentText("Please select a valid file before proceeding.");
            alert.showAndWait();
            return;
        }

        String selectedBank = showBankSelectionDialog();
        if (selectedBank == null) return; // utilizador cancelou

        BankStatementParser parser = switch (selectedBank) {
            case "Crédito Agrícola" -> new CreditoAgricolaParser();
            case "CGD" -> new CGDParser();
            default -> new CreditoAgricolaParser();
        };

        File file = new File(filePath);
        // Kotlin tinha file.nameWithoutExtension — em Java faz-se assim:
        String trimmedName = file.getName().replaceFirst("[.][^.]+$", "");

        try {
            List<Transaction> transactions = parser.parse(file);
            AccountManager newManager = new AccountManager(trimmedName);
            newManager.setBankType(selectedBank);
            newManager.loadTransactions(transactions);
            newManager.saveToFile();
            navigateToDashboard(event, newManager);
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    private String showBankSelectionDialog() {
        Dialog<String> dialog = new Dialog<>();
        dialog.setTitle("Create New Manager");

        ButtonType okButtonType = new ButtonType("OK", ButtonBar.ButtonData.OK_DONE);
        dialog.getDialogPane().getButtonTypes().addAll(okButtonType, ButtonType.CANCEL);

        ComboBox<String> comboBanks = new ComboBox<>();
        comboBanks.setItems(FXCollections.observableArrayList("Crédito Agrícola", "CGD"));
        comboBanks.getSelectionModel().selectFirst();
        comboBanks.setMaxWidth(Double.MAX_VALUE);

        Button btnSettings = new Button("⚙️");
        btnSettings.setOnAction(e -> openSettings());
        // quando o SettingsManager estiver feito, atualiza a lista de bancos aqui

        VBox layout = new VBox(10);
        layout.setPadding(new Insets(20));
        layout.getChildren().addAll(
                new HBox(btnSettings),
                new Label("Select the Bank for this new Manager:"),
                comboBanks
        );

        dialog.getDialogPane().setContent(layout);

        // resultConverter: diz ao Dialog o que devolver quando o utilizador carrega OK ou CANCEL
        dialog.setResultConverter(buttonType -> {
            if (buttonType == okButtonType) return comboBanks.getValue();
            return null;
        });

        return dialog.showAndWait().orElse(null);
    }

    private void openSettings() {
        try {
            FXMLLoader loader = new FXMLLoader(getClass().getResource("/SettingsView.fxml"));
            Parent root = loader.load();
            Stage stage = new Stage();
            stage.setTitle("Bank Reader Settings");
            stage.setScene(new Scene(root));
            stage.initModality(Modality.APPLICATION_MODAL);
            stage.showAndWait();
        } catch (Exception e) {
            System.out.println("Error opening Settings: " + e.getMessage());
        }
    }

    private void navigateToDashboard(ActionEvent event, AccountManager manager) {
        try {
            FXMLLoader loader = new FXMLLoader(getClass().getResource("/Dashboard.fxml"));
            Parent root = loader.load();

            // getController() busca o controller já instanciado pelo FXMLLoader
            DashboardController controller = loader.getController();
            controller.initData(manager);

            Stage stage = (Stage) ((Node) event.getSource()).getScene().getWindow();
            stage.setScene(new Scene(root, 1200, 800));
            stage.setResizable(true);
            stage.centerOnScreen();
            stage.show();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/financemanager/controllers/CreateManagerController.kt
package com.financemanager.controllers

import com.financemanager.IO.importers.BankStatementParser
import com.financemanager.IO.importers.CGDParser
import com.financemanager.IO.importers.CreditoAgricolaParser
import com.financemanager.service.AccountManager
import com.financemanager.service.SettingsManager
import javafx.collections.FXCollections
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.geometry.Insets
import javafx.scene.Node
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.Alert
import javafx.scene.control.Button
import javafx.scene.control.ButtonBar
import javafx.scene.control.ButtonType
import javafx.scene.control.ComboBox
import javafx.scene.control.Dialog
import javafx.scene.control.Label
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.stage.FileChooser
import javafx.stage.Modality
import javafx.stage.Stage
import javafx.util.Callback
import java.io.File
import java.io.IOException

class CreateManagerController {

    @FXML var back_to_menu_button: Button = Button()
    @FXML var import_file_path_button: Button = Button()
    @FXML var import_file_button: Button = Button()
    @FXML var file_name_label: Label = Label()

    var filePath: String = ""

    @FXML
    fun handleBacktoMenu(event: ActionEvent) {
        try {
            val loader = FXMLLoader(javaClass.getResource("/Menu.fxml"))
            val settingsRoot = loader.load<Parent>()

            val menuScene = Scene(settingsRoot)

            val currentStage = (event.source as Node).scene.window as Stage

            currentStage.apply {
                scene = menuScene
                title = "Menu"
                show()
            }
        } catch (e: IOException) {
            println("Error loading FXML: ${e.message}")
        }
    }

    @FXML
    fun handleImportMenu(event: ActionEvent) {
        val fileChooser = FileChooser()
        fileChooser.title = "Import New Manager"

        fileChooser.extensionFilters.addAll(
            FileChooser.ExtensionFilter("Manager Files (*.manager)", "*.csv", "*.xlsx"),
            FileChooser.ExtensionFilter("All Files", "*.*")
        )
        val currentStage = (event.source as Node).scene.window as Stage

        val selectedFile: File? = fileChooser.showOpenDialog(currentStage)
        filePath = selectedFile?.absolutePath ?: ""

        if (selectedFile != null) {
            file_name_label.setText(selectedFile.name)
            println("Selected File: ${selectedFile.absolutePath}")
        } else {
            println("Selection Cancelled")
        }
    }

    @FXML
    fun handleImportManager(event: ActionEvent) {
        if (filePath.isEmpty()) {
            Alert(Alert.AlertType.ERROR).apply {
                title = "Selection Error"
                headerText = null
                contentText = "Please select a valid .manager file before proceeding."
                showAndWait()
            }
            return
        } else {
            val selectedBank = showBankSelectionDialog() ?: return
            val parser = when(selectedBank) {
                "Crédito Agricola" -> CreditoAgricolaParser();
                "CGD" -> CGDParser();
                else -> CreditoAgricolaParser()
            }
            val file: File = File(filePath)
            val trimmedName = file!!.nameWithoutExtension

            try {
                val transactions = parser.parse(file)
                val newManager: AccountManager = AccountManager(trimmedName)
                newManager.bankType = selectedBank
                newManager.loadTransactions(transactions)
                newManager.saveToFile()
                navigateToDashboard(event, newManager)
            } catch (e: Exception) {
                println("Error: ${e.message}")
            }

        }
    }

    private fun showBankSelectionDialog(): String? {
        val dialog = Dialog<String>()
        dialog.title = "Create New Manager"

        val okButtonType = ButtonType("OK", ButtonBar.ButtonData.OK_DONE)
        dialog.dialogPane.buttonTypes.addAll(okButtonType, ButtonType.CANCEL)

        val comboBanks = ComboBox<String>()

        val refreshBanks = {
            //val installed = SettingsManager.getInstance().installedBanks
            val installed = listOf("Crédito Agrícola", "CGD")
            comboBanks.items = FXCollections.observableArrayList(installed)
            comboBanks.selectionModel.selectFirst()
            //comboBanks.items = FXCollections.observableArrayList(*installed)
        }
        refreshBanks()

        val btnSettings = Button("⚙️").apply {
            setOnAction {
                openSettings()
                refreshBanks()
            }
        }

        val layout = VBox(10.0).apply {
            padding = Insets(20.0)
            children.addAll(
                HBox(btnSettings),
                Label("Select the Bank for this new Manager:"),
                comboBanks.apply { maxWidth = Double.MAX_VALUE }
            )
        }

        dialog.dialogPane.content = layout
        dialog.resultConverter = Callback { if (it == okButtonType) comboBanks.value else null }

        return dialog.showAndWait().orElse(null)
    }

    private fun openSettings() {
        try {
            val loader = FXMLLoader(javaClass.getResource("/SettingsView.fxml"))
            val root = loader.load<Parent>()

            val stage = Stage()
            stage.title = "Bank Reader Settings"
            stage.scene = Scene(root)

            stage.initModality(Modality.APPLICATION_MODAL)

            stage.showAndWait()
        } catch (e: Exception) {
            println("Error opening Settings: ${e.message}")
        }
    }

    private fun navigateToDashboard(event: ActionEvent, manager: AccountManager) {
        val loader = FXMLLoader(javaClass.getResource("/Dashboard.fxml"))

        val root = loader.load<Parent>()

        val controller = loader.getController<DashboardController>()

        controller.initData(manager)

        val stage = (event.source as Node).scene.window as Stage
        val scene = Scene(root, 1200.0, 800.0)
        stage.scene = scene

        stage.isResizable = true
        stage.centerOnScreen()
        stage.show()
    }
}